Add EdgeZero canary routing (rollout_pct flag + FNV-1a bucket assignment) (PR 19) (#727)
IABTechLab/trusted-server: docs/superpowers/plans/2026-05-21-pr19-cutover-canary-rollout.md

| Timeout rate     | > 2× baseline            | **Immediate rollback**                 |

> **Note on p95 threshold:** The spec §Cutover paragraph mentions ±10% as the Stage 2 hold-point
> criterion; the threshold table at §Pass/fail thresholds says 15%. These two values are
> inconsistent in the spec. This runbook adopts the threshold table (15%) as the governing
> number because it applies uniformly across all stages. If ops adopts a stricter 10% target
> at Stage 2, update this table accordingly.

---

## Rollback procedure

Rollback is **immediate, no deploy required**.

1. Set `edgezero_rollout_pct = "0"` in the production config store.
   Traffic shifts back to legacy within seconds (next request per Wasm instance).
2. Optionally set `edgezero_enabled = "false"` as belt-and-suspenders.
3. Investigate root cause before re-advancing the canary.
4. Keep the legacy entry point (`legacy_main()`) available until at least one
   full release cycle after reaching 100% with zero regressions.

---

## Monitoring

Fastly real-time stats dashboard. Key signals at each canary stage:

- **Error rate:** `5xx / total_requests` by edge PoP

> - **Latency p95:** use log search for `routing request through EdgeZero path` to identify EdgeZero traffic (`x-edgezero-path` instrumentation header does not exist yet — follow-up task)

Reviewer: Automated review: **P2: Update the duplicated monitoring guidance to match the runbook.**

This plan still embeds stale operational guidance telling operators to use Fastly log search / Fastly log tailing for `routing request through EdgeZero path` to identify EdgeZero traffic. The implemented logger defaults to `Info`, while the route-decision lines are emitted only at `debug!`, and the new runbook explicitly says those lines do not reach production and that no `x-edgezero-path`/real-time-stats branch split exists yet.

Why it matters: if someone uses this plan as the handoff/checklist during cutover, it gives false confidence that production has per-branch route attribution when it does not.

Suggested fix: update or remove this duplicated Monitoring section so it mirrors `docs/internal/EDGEZERO_MIGRATION.md`: production canary verification uses aggregate service metrics only, and route-decision logs are local Viceroy-only with `EDGEZERO_LOG_LEVEL=debug` until a production-safe branch signal is added.

Author: Fixed in 17925a7c. Replaced the stale Monitoring section in the plan with aggregate-metrics-only verification, mirroring `docs/internal/EDGEZERO_MIGRATION.md`: route-decision logs are local Viceroy only (`EDGEZERO_LOG_LEVEL=debug`), production surfaces no per-branch attribution, and no `x-edgezero-path`/real-time-stats split exists yet.

const EDGEZERO_ROLLOUT_PCT_KEY: &str = "edgezero_rollout_pct";
```

- [ ] **Step 2: Implement `read_rollout_pct`**

Add after the `is_edgezero_enabled` function (after line 114):

```rust
/// Reads `edgezero_rollout_pct` from the config store.
///
/// | Config store state              | Return value | Effect                     |
/// |---------------------------------|--------------|----------------------------|
/// | Key absent                      | `0`          | All legacy (safe default)  |
/// | Key present, valid 0–100        | parsed value | Partial or full rollout    |
/// | Key present, invalid            | `0`          | All legacy (safe default)  |
/// | Key read error                  | `0`          | All legacy (safe default)  |
///
fn read_rollout_pct(config_store: &ConfigStoreHandle) -> u8 {
    match config_store.get(EDGEZERO_ROLLOUT_PCT_KEY) {
        Ok(Some(value)) => match parse_rollout_pct(&value) {
            Some(pct) => pct,
            None => {
                log::warn!(
                    "invalid edgezero_rollout_pct value {:?}, defaulting to 0 (legacy path)",
                    value
                );
                0
            }
        },

>         Ok(None) => 100,

Reviewer: Automated review: **P2: This snippet still documents the old fail-open absent-key behavior.**

The table immediately above says an absent `edgezero_rollout_pct` returns `0`, the implementation now returns `0`, and the final invariant says absent/invalid/unreadable must never route traffic to EdgeZero. This embedded code block still says `Ok(None) => 100`, so anyone reusing the implementation plan would reintroduce the exact fail-open behavior this PR just fixed: deleting or omitting the key would trigger full EdgeZero cutover.

Suggested fix: update this arm to return `0` (and preferably include the same debug log/comment as the real implementation), or mark the whole implementation snippet as historical/superseded so it cannot be copied as current guidance.

>         Ok(None) => 100,

Reviewer: 🔧 **wrench — this embedded snippet still has the old fail-open absent-key behavior.**

The table above, the runtime code, and the runbook all say an absent `edgezero_rollout_pct` fails safe to `0`; this snippet still returns `100`. Because this file is an agent handoff, copying it later would reintroduce the exact full-cutover footgun the PR fixed.

```suggestion
        Ok(None) => {
            log::debug!(
                "edgezero_rollout_pct key absent, defaulting to 0 (legacy path — fail safe)"
            );
            0
        }
```

        log::warn!("failed to read edgezero_enabled flag, falling back to legacy path: {e}");
        false
    }) {
        log::debug!("routing request through legacy path (edgezero_enabled=false)");
        legacy_main(req);
        return;
    }

    let rollout_pct = read_rollout_pct(&edgezero_config_store);
    let routing_key = req
        .get_client_ip_addr()
        .map(|ip| ip.to_string())
        .unwrap_or_default();
    let bucket = fnv1a_bucket(&routing_key);

    if canary_routes_to_edgezero(bucket, rollout_pct) {
        log::debug!(
            "routing request through EdgeZero path (bucket={bucket}, rollout_pct={rollout_pct})"
        );
        edgezero_main(req, edgezero_config_store);
    } else {
        log::debug!(
            "routing request through legacy path (bucket={bucket}, rollout_pct={rollout_pct})"
        );
        legacy_main(req);
    }
}
```

- [ ] **Step 4: Run tests, fmt, and clippy**

```bash
cargo test-fastly 2>&1 | tail -20
```

Expected: all tests pass.

```bash
cargo fmt --all -- --check 2>&1 | tail -10
```

Expected: no diff. If there is output, run `cargo fmt --all` to fix, then re-check.

```bash

> cargo clippy -p trusted-server-adapter-fastly --all-targets --all-features -- -D warnings 2>&1 | tail -20

Reviewer: 🔧 **wrench — use the target-matched Fastly clippy gate.**

This host-targeted clippy command conflicts with `CLAUDE.md`'s Fastly gate. Also, the surrounding `| tail`/`| head` validation commands should be removed or wrapped in `set -o pipefail`, otherwise a failed `cargo`/`npm` command can look green because the final pipeline command succeeded.

```suggestion
cargo clippy-fastly
```

---

## Reference

- Spec: `docs/superpowers/specs/2026-03-19-edgezero-migration-design.md` §Cutover plan
- Plan: `docs/superpowers/plans/2026-05-21-pr19-cutover-canary-rollout.md`
- Legacy cleanup: issue [#495](https://github.com/IABTechLab/trusted-server/issues/495)
```

- [ ] **Step 2: Verify `docs/internal/` is covered by Prettier, then check format**

First confirm the Prettier glob recurses into `internal/`:

```bash
cd docs && cat package.json | grep -A3 '"format"'
```

If the glob is `"*.md"` (flat), update it to `"**/*.md"` before proceeding. Then:

```bash
cd docs && npm run format -- --check 2>&1 | tail -10
```

Expected: `All matched files use Prettier code style!` (or no diff output)

If not: `cd docs && npm run format` then re-check.

- [ ] **Step 3: Commit**

```bash
git add docs/internal/EDGEZERO_MIGRATION.md
git commit -m "Add EdgeZero canary rollout ops runbook

Documents the two config store keys, canary progression stages (1% →
10% → 50% → 100%), hold-point durations, pass/fail thresholds, and
instant rollback procedure."
```

---

## Final verification

- [ ] **Run full test suite**

```bash

> cargo test-fastly 2>&1 | tail -20

Reviewer: 🔧 **wrench — don't pipe validation through `tail` without `pipefail`.**

As written, a failing test command can still exit successfully if `tail` succeeds. Keep the command's real exit status visible.

```suggestion
cargo test-fastly
```

```

Expected: all tests pass, including the 10 new canary routing tests.

- [ ] **Run fmt and clippy**

```bash
cargo fmt --all -- --check 2>&1 | tail -10

> cargo clippy -p trusted-server-adapter-fastly --all-targets --all-features -- -D warnings 2>&1 | tail -10

Reviewer: 🔧 **wrench — final verification should use the Fastly clippy alias and preserve exit status.**

This is the PR handoff gate, so it should match the repo's target-specific CI command and not hide failures behind `tail`.

```suggestion
cargo clippy-fastly
```

```

- [ ] **Push branch and open PR against `feature/edgezero-pr18-phase5-verification`**

```bash
git push -u origin feature/edgezero-pr19-cutover-canary
```

Open PR:

- Base branch: `feature/edgezero-pr18-phase5-verification`
- Title: `Add edgezero_rollout_pct canary routing (PR19)`
- Closes: #500

---

## Follow-up items (out of scope for this PR)

- **`x-edgezero-path` diagnostic header**: Add `x-edgezero-path: true` to `edgezero_main` responses so Fastly dashboards and Prettier filters can distinguish EdgeZero vs legacy traffic without log scraping. Track as a separate issue.
- **`docs/README.md` runbook pointer**: Add a line referencing `docs/internal/EDGEZERO_MIGRATION.md` so ops can discover it without grepping.

---

## Invariants to preserve

- **No new dependencies** — FNV-1a is inlined, not a crate import.
- **Fail-safe** — absent, invalid, or unreadable `edgezero_rollout_pct` all produce 0 (all legacy), never 100. A missing or misconfigured key can never accidentally route traffic to EdgeZero; rollout requires an explicit percentage.
- **No handler or core changes** — all changes are confined to `main.rs` entry-point logic and `fastly.toml`.

> - **Sticky routing** — client IP hash is deterministic; the same user always gets the same path at a given `rollout_pct`, preventing split-session bugs.

Reviewer: Automated review: **P2: This invariant still overstates stickiness as per-user.**

The runtime hashes `client_ip`, and the canonical runbook now correctly says routing is sticky per client IP, not per user: NAT/CGNAT users can share a bucket, and a user whose IP changes can be re-bucketed and switch paths. This final invariant still promises that the same user always gets the same path and that split-session bugs are prevented, which contradicts the implemented routing key and the fixed runbook wording.

Suggested fix: align this invariant with the runbook, e.g. `client IP hash is deterministic; the same client IP gets the same path at a given rollout_pct, but NAT sharing and IP changes can re-bucket users.`

> - **Sticky routing** — client IP hash is deterministic; the same user always gets the same path at a given `rollout_pct`, preventing split-session bugs.

Reviewer: 🔧 **wrench — keep the invariant to per-client-IP stickiness.**

The runtime hashes `client_ip`, not a durable user identifier. NAT/CGNAT users can share a bucket, and a user whose IP changes can re-bucket, so the plan should not promise that the same user always gets the same path.

```suggestion
- **Sticky routing** — client IP hash is deterministic; the same client IP gets the same path at a given `rollout_pct`, but NAT sharing and IP changes can re-bucket users.
```

- [ ] **Step 1: Create `docs/internal/` and the runbook file**

```bash
mkdir -p docs/internal
```

Create `docs/internal/EDGEZERO_MIGRATION.md` with exactly this content:

```markdown
# EdgeZero Migration Runbook

Operational reference for the Fastly Compute EdgeZero canary rollout
(issue [#500](https://github.com/IABTechLab/trusted-server/issues/500),
epic [#480](https://github.com/IABTechLab/trusted-server/issues/480)).

---

## Config store keys

Config store name: **`trusted_server_config`** (Fastly service config store)

| Key                    | Type                 | Effect                                                                                                                            |
| ---------------------- | -------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
| `edgezero_enabled`     | `"true"` / `"false"` | Master on/off switch. Set `"false"` to disable EdgeZero entirely, regardless of rollout_pct.                                      |

> | `edgezero_rollout_pct` | `"0"` – `"100"`      | Percentage of traffic (by client IP bucket) routed to EdgeZero. Only read when `edgezero_enabled = "true"`. Key absent = `"100"`. |

Reviewer: Automated review: **P2: The embedded runbook/plan still contradicts the implemented rollout semantics.**

This implementation plan still says absent `edgezero_rollout_pct` means `100`, repeats the `Do NOT delete` warning/setup-ordering precondition below, and still describes routing as sticky per user. The actual code and `docs/internal/EDGEZERO_MIGRATION.md` now say absent/invalid/unreadable all fail safe to legacy and that stickiness is per client IP.

Why it matters: this file is explicitly written as an agentic-worker handoff and includes an embedded ops runbook; if it is used later for follow-up work or cutover prep, it can reintroduce the exact fail-open and stickiness assumptions this PR just fixed.

Suggested fix: update the duplicated sections (background flow, embedded runbook/failure-mode table, production setup precondition, invariants, and sticky-routing wording) to match the final implementation, or replace them with a short pointer to `docs/internal/EDGEZERO_MIGRATION.md` so there is only one source of operational truth.

Author: Fixed in 403de35f. Updated every stale spot in the plan: background flow, the read_rollout_pct doc-table snippet, the rustdoc summary, the embedded fastly.toml snippet, the embedded runbook (config table, failure-modes table, do-not-delete warning, setup-order precondition), the sticky-routing note, and the invariants — all now match the implementation (absent/invalid/unreadable fail safe to 0; sticky per client IP). Also added a pointer marking docs/internal/EDGEZERO_MIGRATION.md as the canonical runbook so the embedded copy cannot silently drift again.

| Create | `docs/internal/EDGEZERO_MIGRATION.md`              | Ops runbook: config keys, canary progression, hold points, thresholds, rollback                                                         |

---

## Background (read this before touching code)

`main.rs:55` already defines `EDGEZERO_ENABLED_KEY = "edgezero_enabled"` and `main.rs:109` implements `is_edgezero_enabled()`.

Current `main()` flow (lines 129–159):

1. Open config store (`open_trusted_server_config_store`) — on error → `legacy_main`
2. Check `edgezero_enabled` flag — if false/err → `legacy_main`
3. If true → `edgezero_main`

**No** `edgezero_rollout_pct` key exists anywhere yet. After this PR the flow becomes:

1. Open config store — on error → `legacy_main`
2. Check `edgezero_enabled` — if false/err → `legacy_main`
3. Read `rollout_pct` from `read_rollout_pct()` (absent / invalid / unreadable all = 0, fail safe to legacy)
4. Compute `bucket = fnv1a_bucket(client_ip_string)`
5. If `bucket < rollout_pct` → `edgezero_main`, else → `legacy_main`

> `ConfigStoreHandle` comes from `edgezero_core::config_store` and cannot be constructed in unit tests (needs Fastly runtime). Test only pure functions.

Reviewer: Automated review: **P2: The implementation plan still says the config-store rollout helper cannot be unit-tested.**

This is now stale and contradicts the PR's final code: `read_rollout_pct` is unit-tested with an in-memory `ConfigStoreHandle` stub in `crates/trusted-server-adapter-fastly/src/main.rs`, and the helper was also renamed from `canary_routes_to_edgezero` to `routes_to_edgezero`. Because this file is explicitly framed as an agentic-worker handoff, leaving the old constraint in place can cause future follow-up work to skip the safety-critical config-store branch tests or copy the outdated helper names.

Suggested fix: either mark this plan as a historical snapshot that must not be reused for implementation, or sync the affected sections with the final PR: remove the claim that `ConfigStoreHandle` cannot be constructed in unit tests, mention the in-memory `ConfigStore` stub coverage for `read_rollout_pct`, and update the old `canary_routes_to_edgezero` references to `routes_to_edgezero`.
